common/io.hpp

```diff
@@ -86,6 +88,26 @@ class Discard : virtual public Writer {
 	}
 };
 
+class StringReader : virtual public Reader {
+private:
+	StreamReader reader_;
+	std::stringstream s_;
+
+public:
+	StringReader(string &str) :
+		s_ {str},
```

Comment:
Any reason why you prefer the `{}` initializers instead of parentheses? You seem to be the only one using them.

Reply:
It's just cause it's what I've learnt, and it is what is recommended to use these days appareantly: https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-list.

 No hard preference here though.